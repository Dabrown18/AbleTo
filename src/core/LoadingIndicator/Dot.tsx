import React, { FunctionComponent } from 'react';
import { StyleSheet } from 'react-native';
import Animated, {
    interpolateColor,
    useAnimatedStyle,
    useSharedValue,
    withDelay,
    withRepeat,
    withSequence,
    withTiming,
} from 'react-native-reanimated';
import styled from 'styled-components';

import useDidMount from '../../hooks/useDidMount';
import colors from '@src/core/colors';

export type DotSize = 'xs' | 'sm' | 'md' | 'lg';

type Props = {
    index: number;
    size: DotSize;
};

const Dot: FunctionComponent<Props> = ({ index, size }) => {
    const progress = useSharedValue(0);

    useDidMount(() => {
        progress.value = withDelay(
            index * 200,
            withRepeat(withSequence(withTiming(1, { duration: 500 }), withTiming(0, { duration: 1000 })), -1)
        );
    });

    const rStyle = useAnimatedStyle(() => {
        const backgroundColor = interpolateColor(progress.value, [0, 1], [INACTIVE_COLOR, ACTIVE_COLOR]);

        return { backgroundColor };
    });

    return <StyledDot size={size} style={rStyle} />;
};

const StyledDot = styled(Animated.View)<{ size: DotSize }>`
    width: ${(props) => styles[props.size].width};
    height: ${(props) => styles[props.size].height};
    background-color: ${colors.gray500};
    margin: 0 5px;
    border-radius: 50px;
`;

const ACTIVE_COLOR = colors.gray500;
const INACTIVE_COLOR = colors.gray400;

const styles = StyleSheet.create({
    xs: {
        width: '4px',
        height: '4px',
    },
    sm: {
        width: '8px',
        height: '8px',
    },
    md: {
        width: '16px',
        height: '16px',
    },
    lg: {
        width: '24px',
        height: '24px',
    },
});

export default Dot;
