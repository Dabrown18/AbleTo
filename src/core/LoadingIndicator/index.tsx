import React from 'react';
import { View, Text } from 'react-native';

const LoadingIndicator = () => {
  return (
    <View>
      <Text>LoadingIndicator</Text>
    </View>
  );
};

export default LoadingIndicator;
