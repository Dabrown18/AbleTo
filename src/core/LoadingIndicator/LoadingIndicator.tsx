import React, { FunctionComponent } from 'react';
import { ViewProps } from 'react-native';
import styled from 'styled-components';

import Dot, { DotSize } from './Dot';
import colors from '@src/core/colors';

const DOTS = Array.from(Array(3));

type Props = {
    fullScreen?: boolean;
    size?: DotSize;
} & Pick<ViewProps, 'style'>;

const LoadingIndicator: FunctionComponent<Props> = ({ fullScreen, size = 'sm', ...rest }) => {
    return (
        <FullScreenCentered fullScreen={fullScreen} accessibilityLabel="Loading indicator">
            <OuterContainer {...rest}>
                {DOTS.map((_, idx) => (
                    <Dot size={size} key={idx} index={idx} />
                ))}
            </OuterContainer>
        </FullScreenCentered>
    );
};

const FullScreenCentered = styled.View<{ fullScreen: boolean | undefined }>`
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    justify-content: center;
    align-items: center;
    z-index: 9001;
    background-color: ${(props) => (props.fullScreen ? colors.grayTransparent700 : 'transparent')};
`;

const OuterContainer = styled.View`
    flex-direction: row;
    background-color: ${colors.white};
    padding: 18px 20px;
    border-radius: 200px;
    align-self: center;
`;

export default LoadingIndicator;
