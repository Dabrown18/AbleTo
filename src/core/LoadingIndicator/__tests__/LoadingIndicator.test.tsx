import { render } from '@testing-library/react-native';
import React from 'react';

import LoadingIndicator from '../LoadingIndicator';

describe('LoadingIndicator snapshot', () => {
    test('Normal state', () => {
        const tree = render(<LoadingIndicator />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('Centered state', () => {
        const tree = render(<LoadingIndicator />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('FullScreen state', () => {
        const tree = render(<LoadingIndicator fullScreen />).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
