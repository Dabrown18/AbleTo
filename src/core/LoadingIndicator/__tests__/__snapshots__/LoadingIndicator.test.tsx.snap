// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIndicator snapshot Centered state 1`] = `
<View
  accessibilityLabel="Loading indicator"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 9001,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "backgroundColor": "#fff",
          "borderRadius": 200,
          "flexDirection": "row",
          "paddingBottom": 18,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 18,
        },
      ]
    }
  >
    <View
      size="sm"
      style={
        Array [
          Object {
            "backgroundColor": "#A6A29E",
            "borderRadius": 50,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 0,
            "width": 8,
          },
          Object {
            "backgroundColor": undefined,
          },
        ]
      }
    />
    <View
      size="sm"
      style={
        Array [
          Object {
            "backgroundColor": "#A6A29E",
            "borderRadius": 50,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 0,
            "width": 8,
          },
          Object {
            "backgroundColor": undefined,
          },
        ]
      }
    />
    <View
      size="sm"
      style={
        Array [
          Object {
            "backgroundColor": "#A6A29E",
            "borderRadius": 50,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 0,
            "width": 8,
          },
          Object {
            "backgroundColor": undefined,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`LoadingIndicator snapshot FullScreen state 1`] = `
<View
  accessibilityLabel="Loading indicator"
  fullScreen={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#5e5c5966",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 9001,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "backgroundColor": "#fff",
          "borderRadius": 200,
          "flexDirection": "row",
          "paddingBottom": 18,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 18,
        },
      ]
    }
  >
    <View
      size="sm"
      style={
        Array [
          Object {
            "backgroundColor": "#A6A29E",
            "borderRadius": 50,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 0,
            "width": 8,
          },
          Object {
            "backgroundColor": undefined,
          },
        ]
      }
    />
    <View
      size="sm"
      style={
        Array [
          Object {
            "backgroundColor": "#A6A29E",
            "borderRadius": 50,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 0,
            "width": 8,
          },
          Object {
            "backgroundColor": undefined,
          },
        ]
      }
    />
    <View
      size="sm"
      style={
        Array [
          Object {
            "backgroundColor": "#A6A29E",
            "borderRadius": 50,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 0,
            "width": 8,
          },
          Object {
            "backgroundColor": undefined,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`LoadingIndicator snapshot Normal state 1`] = `
<View
  accessibilityLabel="Loading indicator"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 9001,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "backgroundColor": "#fff",
          "borderRadius": 200,
          "flexDirection": "row",
          "paddingBottom": 18,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 18,
        },
      ]
    }
  >
    <View
      size="sm"
      style={
        Array [
          Object {
            "backgroundColor": "#A6A29E",
            "borderRadius": 50,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 0,
            "width": 8,
          },
          Object {
            "backgroundColor": undefined,
          },
        ]
      }
    />
    <View
      size="sm"
      style={
        Array [
          Object {
            "backgroundColor": "#A6A29E",
            "borderRadius": 50,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 0,
            "width": 8,
          },
          Object {
            "backgroundColor": undefined,
          },
        ]
      }
    />
    <View
      size="sm"
      style={
        Array [
          Object {
            "backgroundColor": "#A6A29E",
            "borderRadius": 50,
            "height": 8,
            "marginBottom": 0,
            "marginLeft": 5,
            "marginRight": 5,
            "marginTop": 0,
            "width": 8,
          },
          Object {
            "backgroundColor": undefined,
          },
        ]
      }
    />
  </View>
</View>
`;
